zcash_client_sqlite/src/wallet.rs

```diff
@@ -977,15 +979,33 @@ fn parse_block_metadata(
         block_height,
         block_hash,
         sapling_tree_size,
+        if params
+            .activation_height(NetworkUpgrade::Nu5)
+            .iter()
+            .any(|h| h <= &block_height)
+        {
+            orchard_tree_size_opt.ok_or_else(|| {
+                SqliteClientError::from(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!(
+                        "Missing final Orchard note commitment tree size at height: {}",
+                        block_height
+                    ),
+                ))
+            })?
```

Comment:
Isn't this going to immediately fail for existing wallets? Given that we haven't been writing Orchard commitment tree sizes into the wallet since NU5 activation, there will be a long stretch of `None`s between NU5 and whatever height the wallet was last synced to before upgrading to this code.

I currently see three pathways to resolving this issue:
- Make `orchard_tree_size` optional in `BlockMetadata`.
  - This makes the Orchard logic in the rest of the wallet less clean compared to the Sapling logic. The current non-optional approach is definitely what we would have done had we succeeded in deploying Orchard support prior to NU5 activation.
  - The "Orchard wallet birthday" is then effectively "the height at which the `orchard_commitment_tree_size` column of the `blocks` table is first non-null". (We separately need to decide how to handle pool-specific wallet birthdays in general within the wallet structure.)
- Add a required initial online migration scan, which wallets must complete before they do any other wallet operations.
  - We don't have any established API that we could hook this into, so wallets would be entirely reliant on the changelog to notice that they need to start doing this (along with occurrences of this error message, which we could enhance to point developers to the new API).
  - We could maybe detect this issue in the wallet migration code (see below), but it is not an error condition in this pathway (because any online step necessarily occurs after wallet migration, not before).
- Require wallets to force every user to recover from seed in the upgrade that deploys this code.
  - We could do this by adding a check in the wallet migration code, that checks if the most recently-scanned block is at or beyond NU5 activation and has `orchard_commitment_tree_size = NULL`.
  - This would cause expected but maybe unnecessary data loss to users (specifically of data in the `sent_notes` table that is not recoverable from chain).

Reply:
I took a two-pronged approach: I made `orchard_tree_size` and `sapling_tree_size` both optional in `BlockMetadata`, but both required in `ScannedBlock`.